File: blink-macos/BlinkCore/Sources/BlinkCore/BlinkEngine.swift
import Foundation

/// Central orchestrator for Blink's break reminder logic.
///
/// Receives input events and a 1-second tick from the platform layer.
/// Produces callbacks when the UI needs to show a break, toast, or update.
///
/// All decision logic lives here — the platform layer is just plumbing.
public final class BlinkEngine {

    // MARK: - Callbacks (set by the platform layer)

    /// Show the fullscreen break overlay. `breakNumber` is for walk suggestion (≥4).
    public var onShowBreak: ((_ breakNumber: Int) -> Void)?

    /// Show a toast when the timer extends (e.g., "Active typing — +10 min").
    public var onShowExtendToast: ((_ reason: String) -> Void)?

    /// Timer countdown updated — drives the menu bar display.
    public var onTimerUpdate: ((_ remaining: TimeInterval, _ total: TimeInterval) -> Void)?

    /// State changed — drives the menu bar badge (Working/Away/Meeting/Break).
    public var onStateChange: ((_ state: DisplayState) -> Void)?

    /// Log a message (debug toast, console, etc.)
    public var onLog: ((_ message: String) -> Void)?

    public enum DisplayState: String {
        case working, away, meeting, onBreak
    }

    // MARK: - Configuration

    /// User's sensitivity setting (0.4–0.9). Drives maxExtensions and score threshold.
    public var sensitivity: Double = 0.7 {
        didSet {
            config = FlowConfig.config(forSensitivity: sensitivity)
            decisionEngine.sensitivity = sensitivity
        }
    }

    /// Wall-clock cap: force a break after this many seconds since the last break,
    /// regardless of idle pauses or extensions. User-configurable. Default 40 min.
    public var maxWallClockSeconds: TimeInterval = 2400

    // MARK: - Sub-engines

    private let timer = TimerStateMachine()
    private let stateMachine = FlowStateMachine()
    private let decisionEngine = BreakDecisionEngine()
    private let complianceTracker = BreakComplianceTracker()
    private var config: FlowConfig

    // MARK: - Activity timestamps (single source of truth for idle)

    private var lastKeystrokeTime = Date()
    private var lastClickTime = Date()
    private var lastScrollTime = Date()
    private var lastAppSwitchTime = Date()

    private var lastActivityTime: Date {
        max(lastKeystrokeTime, lastClickTime, lastScrollTime, lastAppSwitchTime)
    }

    private var idle: TimeInterval {
        Date().timeIntervalSince(lastActivityTime)
    }

    private var keyboardIdle: TimeInterval {
        Date().timeIntervalSince(lastKeystrokeTime)
    }

    // MARK: - State

    private var breakPending = false
    private var breakPendingSince: Date?
    private var consecutiveBreaks: Int = 0
    private var lastBreakEndedAt: Date?
    private var engineStartTime = Date()
    private var isOnBreak = false
    private var micActive = false
    private var cameraActive = false
    private var videoPlaying = false

    // MARK: - Constants

    private let graceSeconds: TimeInterval = 60
    private let idleThreshold: TimeInterval = 180
    private let courtesyWaitMax: TimeInterval = 10
    private let courtesyGap: TimeInterval = 3

    // MARK: - Init

    public init() {
        self.config = FlowConfig.config(forSensitivity: sensitivity)

        timer.onBreakDue = { [weak self] in
            self?.handleBreakDue()
        }
    }

    // MARK: - Input (called by platform layer)

    public func recordKeystroke() {
        lastKeystrokeTime = Date()
        decisionEngine.recordKeystroke()
    }

    public func recordClick() {
        lastClickTime = Date()
        decisionEngine.recordClick()
    }

    public func recordScroll() {
        lastScrollTime = Date()
        decisionEngine.recordScroll()
    }

    public func recordAppSwitch(bundleID: String) {
        lastAppSwitchTime = Date()
        decisionEngine.recordAppSwitch(bundleID: bundleID)
    }

    public func setMicActive(_ active: Bool) {
        micActive = active
    }

    public func setCameraActive(_ active: Bool) {
        cameraActive = active
    }

    public func setVideoPlaying(_ playing: Bool) {
        if playing != videoPlaying {
            videoPlaying = playing
            if playing {
                // Video playback = not close-up screen work → reset timer
                timer.resetAfterBreak()
                decisionEngine.resetAll()
                onTimerUpdate?(timer.remainingSeconds, timer.timerDuration)
            }
        }
    }

    public func wakeFromSleep() {
        // Idle is automatically correct from timestamps.
        // Force an immediate check so we don't wait for the next tick.
        tickInternal()
    }

    // MARK: - Break response (called by platform layer)

    public func userTookBreak() {
        consecutiveBreaks += 1
        complianceTracker.breakTaken(at: Date(), idleDuration: 20)
        finishBreak()
    }

    public func userSkippedBreak() {
        complianceTracker.breakDismissed(at: Date())
        finishBreak()
    }

    public func userSnoozed(minutes: Int) {
        isOnBreak = false
        timer.reset(duration: TimeInterval(minutes * 60))
        onTimerUpdate?(timer.remainingSeconds, timer.timerDuration)
        onStateChange?(.working)
    }

    // MARK: - Tick (called every 1 second by platform layer)

    public func tick() {
        tickInternal()
    }

    private func tickInternal() {
        // 1. Pending break delivery (10s courtesy wait for keyboard gap)
        if breakPending {
            let waited = Date().timeIntervalSince(breakPendingSince ?? Date())
            if waited >= courtesyWaitMax || keyboardIdle >= courtesyGap {
                breakPending = false
                breakPendingSince = nil
                isOnBreak = true
                complianceTracker.breakPrompted(at: Date(), flowState: .normal, flowScore: 0)
                onShowBreak?(consecutiveBreaks + 1)
                onStateChange?(.onBreak)
            }
            return
        }

        // Don't tick while break overlay is showing
        if isOnBreak { return }

        // 2. Wall-clock safety cap
        let sinceLastBreak = Date().timeIntervalSince(lastBreakEndedAt ?? engineStartTime)
        if sinceLastBreak >= maxWallClockSeconds {
            breakPending = true
            breakPendingSince = Date()
            return
        }

        // 3. Grace period (suppress idle detection for 60s after break)
        let inGrace = lastBreakEndedAt.map { Date().timeIntervalSince($0) < graceSeconds } ?? false

        // 4. Idle/meeting detection
        let currentIdle = inGrace ? 0 : idle
        stateMachine.tick(
            flowScore: 0,
            secondsSinceLastInput: currentIdle,
            isMicActive: micActive,
            isCameraActive: cameraActive,
            now: Date().timeIntervalSinceReferenceDate
        )

        let newState = displayState(from: stateMachine.state)
        onStateChange?(newState)

        // 5. Idle reset (eyes already rested)
        if currentIdle >= idleThreshold && !inGrace {
            consecutiveBreaks = 0
            decisionEngine.resetAll()
            timer.resetAfterBreak()
            onTimerUpdate?(timer.remainingSeconds, timer.timerDuration)
            return
        }

        // 6. Video playing — already handled in setVideoPlaying, but skip timer tick
        if videoPlaying { return }

        // 7. Tick timer (pauses if idle/meeting)
        let timerState: FlowState = (stateMachine.state == .idle || stateMachine.state == .meeting)
            ? stateMachine.state : .normal
        timer.tick(flowState: timerState, deltaSeconds: 1.0)
        onTimerUpdate?(timer.remainingSeconds, timer.timerDuration)
    }

    // MARK: - Break decision

    private func handleBreakDue() {
        let decision = decisionEngine.decide(maxExtensions: config.maxExtensions)

        switch decision {
        case .extend(_, let reason):
            decisionEngine.resetWindow()
            timer.reset(duration: 10 * 60)
            onShowExtendToast?(reason)
            onTimerUpdate?(timer.remainingSeconds, timer.timerDuration)

        case .showBreak:
            decisionEngine.resetWindow()
            breakPending = true
            breakPendingSince = Date()
        }
    }

    private func finishBreak() {
        lastBreakEndedAt = Date()
        isOnBreak = false
        decisionEngine.resetWindow()
        timer.resetAfterBreak()
        onTimerUpdate?(timer.remainingSeconds, timer.timerDuration)
        onStateChange?(.working)
    }

    private func displayState(from state: FlowState) -> DisplayState {
        switch state {
        case .idle: return .away
        case .meeting: return .meeting
        case .breakPrompted: return .onBreak
        default: return .working
        }
    }

    // MARK: - Public state accessors

    public var remainingSeconds: TimeInterval { timer.remainingSeconds }
    public var timerDuration: TimeInterval { timer.timerDuration }
    public var currentBreakStreak: Int { consecutiveBreaks }
    public var currentState: DisplayState { displayState(from: stateMachine.state) }

    /// Compliance tracker — expose for persistence in the app layer.
    public var compliance: BreakComplianceTracker { complianceTracker }
}

File: blink-windows/src/Blink.Core/BlinkEngine.cs
using Blink.Core.Compliance;
using Blink.Core.FlowDetection;
using Blink.Core.Timer;

namespace Blink.Core;

/// <summary>
/// Central orchestrator for Blink's break reminder logic.
/// Receives input events and a 1-second tick from the platform layer.
/// Produces callbacks when the UI needs to show a break, toast, or update.
/// Port of Swift BlinkCore.BlinkEngine.
/// </summary>
public sealed class BlinkEngine
{
    // Callbacks
    public Action<int>? OnShowBreak;              // breakNumber
    public Action<string>? OnShowExtendToast;     // reason
    public Action<double, double>? OnTimerUpdate; // remaining, total
    public Action<DisplayState>? OnStateChange;

    public enum DisplayState { Working, Away, Meeting, OnBreak }

    // Config
    /// <summary>
    /// User's sensitivity setting (0.4–0.9). Required at construction —
    /// BlinkCore has no default; the canonical default lives in the consumer.
    /// </summary>
    public double Sensitivity
    {
        get => _sensitivity;
        set
        {
            _sensitivity = value;
            _config = FlowConfig.ForSensitivity(value);
            _decisionEngine.Sensitivity = value;
        }
    }
    private double _sensitivity;

    /// <summary>
    /// Wall-clock cap: force a break after this many seconds since the last
    /// break, regardless of idle pauses or extensions. Default 40 min.
    /// </summary>
    public double MaxWallClockSeconds { get; set; } = 2400;

    // Sub-engines
    private readonly TimerStateMachine _timer = new();
    private readonly FlowStateMachine _stateMachine = new();
    private readonly BreakDecisionEngine _decisionEngine;
    private readonly BreakComplianceTracker _complianceTracker = new();
    private FlowConfig _config;

    /// <summary>Decision engine — exposed for spot-check / tests.</summary>
    public BreakDecisionEngine DecisionEngine => _decisionEngine;

    // Activity timestamps (nullable — set on first event)
    private DateTime? _lastKeystrokeTime;
    private DateTime? _lastClickTime;
    private DateTime? _lastScrollTime;
    private DateTime? _lastAppSwitchTime;

    private DateTime LastActivityTime =>
        new[] { _lastKeystrokeTime, _lastClickTime, _lastScrollTime, _lastAppSwitchTime }
            .Where(t => t.HasValue)
            .Select(t => t!.Value)
            .DefaultIfEmpty(_engineStartTime ?? Now)
            .Max();

    // State
    private bool _breakPending;
    private DateTime? _breakPendingSince;
    private int _consecutiveBreaks;
    private DateTime? _lastBreakEndedAt;
    private DateTime? _engineStartTime;
    private bool _isOnBreak;
    private bool _micActive;
    private bool _cameraActive;
    private bool _videoPlaying;

    // Constants
    private const double GraceSeconds = 20;
    private const double IdleThreshold = 180;
    private const double CourtesyWaitMax = 10;
    private const double CourtesyGap = 3;

    /// <summary>
    /// Simulated clock for testing. When set, all time reads use this instead
    /// of DateTime.UtcNow.
    /// </summary>
    public DateTime? SimulatedNow { get; set; }

    private DateTime Now => SimulatedNow ?? DateTime.UtcNow;

    private static double ToUnix(DateTime t) =>
        ((DateTimeOffset)t).ToUnixTimeMilliseconds() / 1000.0;

    public BlinkEngine(double sensitivity)
    {
        _sensitivity = sensitivity;
        _decisionEngine = new BreakDecisionEngine(sensitivity);
        _config = FlowConfig.ForSensitivity(sensitivity);
        _timer.OnBreakDue += HandleBreakDue;
    }

    // MARK: - Input

    public void RecordKeystroke()
    {
        var t = Now;
        _lastKeystrokeTime = t;
        _engineStartTime ??= t;
        _decisionEngine.RecordKeystroke();
    }

    public void RecordClick()
    {
        var t = Now;
        _lastClickTime = t;
        _engineStartTime ??= t;
        _decisionEngine.RecordClick();
    }

    public void RecordScroll()
    {
        var t = Now;
        _lastScrollTime = t;
        _engineStartTime ??= t;
        _decisionEngine.RecordScroll();
    }

    public void RecordAppSwitch(string processName)
    {
        var t = Now;
        _lastAppSwitchTime = t;
        _engineStartTime ??= t;
        _decisionEngine.RecordAppSwitch(processName, ToUnix(t));
    }

    /// <summary>
    /// Seed the currently-frontmost app at session start, without incrementing
    /// the switch counter. Without this, app-switch events only fire on
    /// CHANGES, so a user who launches Blink into a creative app and never
    /// switches produces zero dwell data — silently losing the creative bonus.
    /// </summary>
    public void SetCurrentApp(string processName)
    {
        var t = Now;
        _engineStartTime ??= t;
        _decisionEngine.SetCurrentApp(processName, ToUnix(t));
    }

    public void SetMicActive(bool active)
    {
        if (active && !_micActive)
        {
            // Meeting started → user isn't staring at code. Treat as eye rest.
            _consecutiveBreaks = 0;
            ResetTimerState();
        }
        _micActive = active;
    }

    public void SetCameraActive(bool active) => _cameraActive = active;

    public void SetVideoPlaying(bool playing)
    {
        if (playing != _videoPlaying)
        {
            _videoPlaying = playing;
            if (playing)
            {
                // Video playback = not close-up screen work → treat as eye rest.
                _consecutiveBreaks = 0;
                ResetTimerState();
            }
        }
    }

    public void WakeFromSleep()
    {
        // Sleep means the user wasn't looking at the screen. Treat as eye rest.
        _consecutiveBreaks = 0;
        ResetTimerState();
    }

    // MARK: - Break response

    public void UserTookBreak()
    {
        _consecutiveBreaks++;
        _complianceTracker.BreakTaken(Now, 20);
        ResetTimerState();
        OnStateChange?.Invoke(DisplayState.Working);
    }

    public void UserSkippedBreak()
    {
        _complianceTracker.BreakDismissed(Now);
        ResetTimerState();
        OnStateChange?.Invoke(DisplayState.Working);
    }

    public void UserSnoozed(int minutes)
    {
        _isOnBreak = false;
        _timer.Reset(minutes * 60);
        OnTimerUpdate?.Invoke(_timer.RemainingSeconds, _timer.TimerDuration);
        OnStateChange?.Invoke(DisplayState.Working);
    }

    // MARK: - Tick

    public void Tick()
    {
        var currentTime = Now;
        _engineStartTime ??= currentTime;

        // 1. Pending break delivery (courtesy wait for keyboard gap).
        //    Don't deliver during a meeting — hold pending until mic inactive.
        if (_breakPending)
        {
            if (!_micActive)
            {
                var waited = (currentTime - (_breakPendingSince ?? currentTime)).TotalSeconds;
                var kbIdle = _lastKeystrokeTime.HasValue
                    ? (currentTime - _lastKeystrokeTime.Value).TotalSeconds : 999;
                if (waited >= CourtesyWaitMax || kbIdle >= CourtesyGap)
                {
                    _breakPending = false;
                    _breakPendingSince = null;
                    _isOnBreak = true;
                    _complianceTracker.BreakPrompted(currentTime, FlowState.Normal, 0);
                    OnShowBreak?.Invoke(_consecutiveBreaks + 1);
                    OnStateChange?.Invoke(DisplayState.OnBreak);
                }
            }
            return;
        }

        if (_isOnBreak) return;

        // 2. Grace + idle measurement (grace zeroes idle to prevent flapping).
        var inGrace = _lastBreakEndedAt.HasValue &&
            (currentTime - _lastBreakEndedAt.Value).TotalSeconds < GraceSeconds;
        var currentIdle = inGrace ? 0 : (currentTime - LastActivityTime).TotalSeconds;

        // 3. Tick state machine + publish display state.
        var nowRef = ToUnix(currentTime);
        _stateMachine.Tick(0, currentIdle, _micActive, _cameraActive, nowRef);
        OnStateChange?.Invoke(MapState(_stateMachine.State));

        // 4. Idle reset (eyes already rested). MUST come before the wall-clock
        //    cap — otherwise extended idle (e.g. wake from sleep) trips the cap
        //    before the reset can clear _lastBreakEndedAt, queuing a phantom break.
        if (currentIdle >= IdleThreshold)
        {
            _consecutiveBreaks = 0;
            ResetTimerState();
            return;
        }

        // 5. Wall-clock safety cap — counts active work, skipped during meetings.
        if (!_micActive)
        {
            var sinceLastBreak =
                (currentTime - (_lastBreakEndedAt ?? _engineStartTime ?? currentTime)).TotalSeconds;
            if (sinceLastBreak >= MaxWallClockSeconds)
            {
                _breakPending = true;
                _breakPendingSince = currentTime;
                return;
            }
        }

        // 6. Video playing — already handled in SetVideoPlaying, but skip tick.
        if (_videoPlaying) return;

        // 7. Feed decision engine (tracks window duration).
        _decisionEngine.Tick(nowRef);

        // 8. Tick timer (pauses if idle/meeting).
        var timerState = _stateMachine.State is FlowState.Idle or FlowState.Meeting
            ? _stateMachine.State : FlowState.Normal;
        _timer.Tick(timerState, 1.0);
        OnTimerUpdate?.Invoke(_timer.RemainingSeconds, _timer.TimerDuration);
    }

    // MARK: - Break decision

    private void HandleBreakDue()
    {
        var t = ToUnix(Now);
        var decision = _decisionEngine.Decide(_config.MaxExtensions, t);

        switch (decision)
        {
            case BreakDecision.Extend ext:
                _decisionEngine.ResetWindow(t);
                _timer.Reset(10 * 60);
                OnShowExtendToast?.Invoke(ext.Reason);
                OnTimerUpdate?.Invoke(_timer.RemainingSeconds, _timer.TimerDuration);
                break;

            case BreakDecision.ShowBreak:
                _decisionEngine.ResetWindow(t);
                _breakPending = true;
                _breakPendingSince = Now;
                break;
        }
    }

    /// <summary>
    /// Shared "user just rested" reset. Restarts the wall-clock cap (sets
    /// _lastBreakEndedAt = now) and clears any pending break / on-break state.
    /// Caller owns _consecutiveBreaks and any state-change callback.
    /// </summary>
    private void ResetTimerState()
    {
        _lastBreakEndedAt = Now;
        _breakPending = false;
        _breakPendingSince = null;
        _isOnBreak = false;
        _decisionEngine.ResetAll(ToUnix(Now));
        _timer.ResetAfterBreak();
        OnTimerUpdate?.Invoke(_timer.RemainingSeconds, _timer.TimerDuration);
    }

    private static DisplayState MapState(FlowState state) => state switch
    {
        FlowState.Idle => DisplayState.Away,
        FlowState.Meeting => DisplayState.Meeting,
        FlowState.BreakPrompted => DisplayState.OnBreak,
        _ => DisplayState.Working
    };

    // Public accessors
    public double RemainingSeconds => _timer.RemainingSeconds;
    public double TimerDuration => _timer.TimerDuration;
    public int CurrentBreakStreak => _consecutiveBreaks;
    public DisplayState CurrentState => MapState(_stateMachine.State);
    public BreakComplianceTracker Compliance => _complianceTracker;

    /// <summary>Snapshot of current flow score and signals without affecting state.</summary>
    public BreakDecisionEngine.SpotCheckResult SpotCheckFlow() => _decisionEngine.SpotCheck();
}
